pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/SinkBase.java

```diff
@@ -0,0 +1,258 @@
+            @ApiResponse(code = 404, message = "The function doesn't exist")
+    })
+    @Path("/{tenant}/{namespace}/{sinkName}/{instanceId}/status")
+    public Response getSinkInstanceStatus(final @PathParam("tenant") String tenant,
+                                          final @PathParam("namespace") String namespace,
+                                          final @PathParam("sinkName") String sinkName,
+                                          final @PathParam("instanceId") String instanceId) throws IOException {
+        return functions.getFunctionInstanceStatus(
+            tenant, namespace, sinkName, instanceId, uri.getRequestUri());
+    }
+
+    @GET
+    @ApiOperation(
+            value = "Displays the status of a Pulsar Sink running in cluster mode",
+            response = FunctionStatus.class
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Invalid request"),
+            @ApiResponse(code = 403, message = "The requester doesn't have admin permissions")
+    })
+    @Path("/{tenant}/{namespace}/{sinkName}/status")
+    public Response getSinkStatus(final @PathParam("tenant") String tenant,
+                                  final @PathParam("namespace") String namespace,
+                                  final @PathParam("sinkName") String sinkName) throws IOException {
+        return functions.getFunctionStatus(tenant, namespace, sinkName, uri.getRequestUri());
+    }
+
+    @GET
+    @ApiOperation(
+            value = "Lists all Pulsar Sinks currently deployed in a given namespace",
+            response = String.class,
+            responseContainer = "Collection"
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Invalid request"),
+            @ApiResponse(code = 403, message = "The requester doesn't have admin permissions")
+    })
+    @Path("/{tenant}/{namespace}")
+    public Response listSinks(final @PathParam("tenant") String tenant,
+                              final @PathParam("namespace") String namespace) {
+        return functions.listFunctions(tenant, namespace);
+
+    }
+
+    @POST
+    @ApiOperation(value = "Restart sink instance", response = Void.class)
+    @ApiResponses(value = { @ApiResponse(code = 400, message = "Invalid request"),
+            @ApiResponse(code = 404, message = "The function does not exist"),
+            @ApiResponse(code = 500, message = "Internal server error") })
+    @Path("/{tenant}/{namespace}/{sinkName}/{instanceId}/restart")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response restartSink(final @PathParam("tenant") String tenant,
+            final @PathParam("namespace") String namespace, final @PathParam("sinkName") String sinkName,
+            final @PathParam("instanceId") String instanceId) {
+        return functions.restartFunctionInstance(tenant, namespace, sinkName, instanceId, uri.getRequestUri());
```

Comment:
sink and source endpoints are just delegating calls to function, how we can prevent people are not using sink endpoints to start sources or other functions?

for example, if a sink is t/ns/name, people can end up using source endpoints or function endpoint to restart sink, right?

Reply:
As i replied to @jerrypeng comments earlier, this is just part 2 of a long phase of getting rid of commonality between sources/sinks. The checks of listing/getinfo/deleting will come in later pr.